### ISSUE TITLE
Clear separation between data & user configuration

### ISSUE THREAD

**Issue by frenck** [home-assistant/architecture#283]
## Context

Currently, we use the `.storage` folder to store all "generated" data in. Several registry's and some integrations, already use the folder. However, a lot of integrations and/or functionality provided by Home Assistant, do not.

Resulting in a mix of user configuration and "other" data.

Some examples include:
- `/config/.ios.conf`
- `/config/.uuid`
- `/config/.spotify-token-cache`
- `/config/.HA_VERSION`
- `/config/OZW-Log.txt`
- `/config/home-assistant.log`
- `/config/home-assistant_v2.db`
- `/config/.cloud/`

## Proposal

From the context given above, a clear pattern is visible, besides, the manual user configuration (YAML), we have to deal with "data" & "logs".

To provide a clear separation between manual user configuration (YAML), stored data & logs, the proposal is to create a clear separation between all those categories of files.

Suggested structure:

- `/config` -> General user configuration.
- `/config/.logs` -> Log files.
- `/config/.storage` -> Data storage.

## Consequences

- We no longer accept integrations or other changes to the Home Assistant core that add or modify the user configuration space. Unless the change contributes to the suggested proposal.
- Identify and migrate these cases in the Home Assistant core.
- Identify and migrate these cases in the current integrations that ship with Home Assistant.

## Variants

There are some variants possible. Things to consider:

- The `/config/.logs` folder is a dotted folder, this might cause it to be "too" hidden. Making this just `/config/logs` is a consideration.
- The Home Assistant database might deserve its own data folders (e.g., `/config/.db`). However, it is still just data as well.
- We should consider splitting core & integration data (which would add another storage folder, or, a subfolder structure in our `.storage` folder.
- From a true user config perspective, the `.storage` folder has the same flaw as the context in the above proposal since it contains both user configuration (stored from UI) as data (e.g., tokens). We should consider spitting that as well.

## Higher impact variation

Adding a split between user configuration, data & logs on even a higher level, taking them out of the user configuration folder completely. In the above context/examples it would look something like this:

- `/config/`
- `/logs/`
- `/storage/` or `/data/`

This impact of this change is much higher since it directly hits the users configuration on all installation methods we provide. Nevertheless, it might be the preferred end goal.

## Notes

I realize this proposal is pretty large, and hard to implement. However, this proposal might be split up into several smaller ADR's to make progress. This depends a little on the general acceptance and comments as a result of this proposal.

**Comment by michaelarnauts** [comment]
IMO, config and storage is the same, but splitting the logs from the config would be nice.

**Comment by frenck** [comment]
@michaelarnauts That comment makes sense, I guess. Which makes me realize that I kinda miss some clarity in my opening post. I'll adjust a bit.

Mainly, the goal of this proposal is to separate user manual configurations (YAML), from the rest. However, you are right, the storage folder contains both data & user configurations (user configurations made from the UI). From that perspective, even more splitting would be needed.

**Comment by Santobert** [comment]
I see some other generated data files that are actually like manual configuration files, e.g. `.storage/core.area_registry`, that is based on user input. The files written during config_flow are also somehow generated manually because they are based on user input.

I would like to have an option to back up all these generated files based on user input as described here:
https://www.home-assistant.io/docs/ecosystem/backup/backup_github/

Since we have config_flow I miss an option to securely backup the whole configuration because there are integrations that are completely configured by config_flow (ESPHome). These doesn't show up in configuration.yaml and wouldn't be part of the backup. Furthermore I don't want to back up the `.storage` folder because it contains sensitive data. That's the reason why we need an option to separate passwords and sensitive data from these configuration files. Something like this:

TL;DR: There are two types of generated data files. These that are easily reproducible because they are based on machine-readable data and those that are based on user input. The seconds ones contain passwords and other sensitive data that we should separate in order to have a secure way to secure the entire configuration.

What do you guys think about that?

**Comment by Jc2k** [comment]
When i first started using HA I wanted to bake my config into my customised Docker image and deploy them together. I would lint this build artifact as a whole before deploying it. This was problematic because there was no seperation of configuration, state (.storage) and logs. A lot of files seemed to be found relative to where my configuration.yaml lived. So i had to have a bash script that copied the config from the docker image into the state directory along side .storage etc.

These days I embrace config entries as much as possible and manage my UI setup through the UI. So i just want to backup my state directory and avoid over complicated Git managed config. But past me would have loved this proposal.

Linux people have fretted about this problem for a long time and the desktop people tend to follow the [XDG Base Directory Spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html). Not proposing that we follow it, but it's an interesting data point. That really defines 4 data types:

 * config
 * cache (this is explicitly stuff that you don't need to backup)
 * runtime (stuff like unix sockets, named pipes, pid files, completely ephemeral)
 * data

For some of these you can actually set a search patch, so it could look for configuration.yml in multiple directories.

Some people have [asked](https://lists.freedesktop.org/pipermail/xdg/2012-December/012598.html) for a state type, but that has never happened.

The spec is a bit vague about what data is. When i see the default directory is `~/.local/share` i assumed it wasn't for data like out sqlite databases. But there seems to be a consensus that thats what is meant. Debian have a further [wiki page](https://wiki.debian.org/XDGBaseDirectorySpecification) with a helpful table:

  | DATA | CONFIG | STATE | CACHE | RUNTIME
-- | -- | -- | -- | -- | --
sync across machines? | yes? | yes | no | no | no
manage in VCS (Git/SVN)? | no | yes | no | no | no
should be backed up? | yes | yes | yes | no | no
can live in tmpfs? | no | no | no | yes | yes?
contains much data? | yes | no | no | yes | no

Maybe we need a similar table.

**Comment by Jc2k** [comment]
@Santobert 

RE: .storage - i think i disagree about further splitting the data in the files in this folder. With configuration.yml you are in control of what is a secret. So i can choose to not put my devices IP addresses in Git when i share a config on GitHub. But .storage is meant for use with the UI. I can't tell it what I want to be private. If we are conservative about what is private or sensitive them what we end up is most of that data is in .storage as it is now. Your example about areas - if i have areas that I want to keep private (maybe family members names for bedrooms for example) then they cannot go in this non-sensitive directory either.

We probably need to come up with more documented backup strategies as more things use .storage. Maybe a GPG offsite one?

**Comment by Santobert** [comment]
@Jc2k That are some interesting things you came up with. I like the 4 types of data, you introduced.

To make my points more clear:
In your table is the line _manage in VCS_ which says yes to configuration. This is exactly what I'm doing. So my .gitignore includes some parts of .storage as well as all yaml files. I like the idea of versioning my configuration. This repository is private, I don't want to share any of these information. In my opinion passwords are still more private. I don't even want to replicate them. All passwords are tagged with [CONF_PASSWORD](https://github.com/home-assistant/home-assistant/blob/1d3f2d20d2eb2d12d02cde93d7d79e66f8157a1c/homeassistant/const.py#L109) during config_flow so we should be able to identify them. I would like to have all these even more sensitive data within one file that could be excluded by .gitignore.

For the rest I agree to you. There could be different opinions about how sensitive some configurations are. But from my point of view there is a huge difference between normal configuration (areas, devices, etc.) and credentials.

**Comment by Santobert** [comment]
I like the table you introduced. So I've extended it with the column _credentials_. The cells are always no here, because I don't want to recreate the credentials at all but they have to be persistent on that machine.

| | DATA | CONFIG | CREDENTIALS | STATE | CACHE | RUNTIME
-- | -- | -- | -- | -- | -- | --
sync across machines? | yes? | yes | no | no | no | no
manage in VCS (Git/SVN)? | no | yes | no | no | no | no
should be backed up? | yes | yes | no | yes | no | no
can live in tmpfs? | no | no | no | no | yes | yes?
contains much data? | yes | no | no | no | yes | no

**Comment by Jc2k** [comment]
CONF_PASSWORD is probably not enough. I know of config entries that have private encryption keys in them (the password that a user entered is used as part of a handshake to set up mutually authenticated key based encryption). I think there are oauth2 tokens in others.

I don’t think of anything in .storage as config, it’s opaque data managed through the UI. Would you commit an SQLite database to git if that’s the format that was chosen for .storage? (It’s worth noting that those files are JSON and not YAML in part because users aren’t supposed to treat them like pets as with configuration.yml).

**Comment by Santobert** [comment]
> CONF_PASSWORD is probably not enough. I know of config entries that have private encryption keys in them (the password that a user entered is used as part of a handshake to set up mutually authenticated key based encryption). I think there are oauth2 tokens in others.

I guess you have a good point here.

> I don’t think of anything in .storage as config, it’s opaque data managed through the UI. Would you commit an SQLite database to git if that’s the format that was chosen for .storage? (It’s worth noting that those files are JSON and not YAML in part because users aren’t supposed to treat them like pets as with configuration.yml).

I disagree here. Even though I've configured the system through the UI the resulting information is still part of the config. I'm glad they are in JSON and not in sqlite :smiley: People experiment a lot with the user interface and configuration.yaml. They may want to undo their experiments some times, which Git allows and Sqlite does not.

**Comment by Jc2k** [comment]
You are advocating for making a backup before experimenting in the UI (I agree!), but using git (ala etckeeper) instead of a dedicated backup tool. But you don’t trust that backup tool to hold your passwords. Could you use git-crypt instead of complicating home-assistant?

**Comment by Santobert** [comment]
This would be an option of course. I just want to leave my opinion here. That there is no difference between configuration in YAML and UI and the resulting files and (and that's just nice-to-have) that I would like to have something like secrets.yaml for the config files in .storage. The advantage of this would be the possibility to write the entire configuration into a VCS.

**Comment by ties** [comment]
_Given_ that files in `.storage` are mean to be managed by the system instead of by the users, perhaps types that are secret could be stored encrypted by default?

There are three main ways to get there:
  * Encrypt secrets entries in file, decrypt all secrets entries when loading (design used by ansible vault)
  * Store ciphertext for field, query another component for decryption when needing to decrypt that field (called ['transit' in hashicorp vault](https://www.vaultproject.io/api/secret/transit/index.html#decrypt-data))
  * Store ciphertext for field in secure storage, query secure storage for entries under a key (design of keychain on OS X)

The latter two approaches can be combined. You need to point to the provided that decrypts the entry. This can be a string that contains, for example: the key used, encryption used, provider used, entry name, …

**Comment by frenck** [comment]
Solved in #366
